quilt3distribute/dataset.py

```diff
@@ -252,30 +307,99 @@ def distribute(
 
                     # Update values to the logical key as they are set
                     for i, val in enumerate(v_ds.data[col].values):
+                        # Fully resolve the path
                         pk = Path(val).expanduser().resolve()
+
+                        # Just using val.name could result in files that shouldn't be grouped being grouped
+                        # Example column:
+                        # SourceReadpath
+                        # a/0.tiff
+                        # a/1.tiff
+                        # b/0.tiff
+                        # b/1.tiff
+                        # Even though there are four files, this would result in both a/0.tiff and b/0.tiff, and,
+                        # a/1.tiff and b/1.tiff being grouped together. To solve this we can prepend a the first couple
+                        # of characters from a hash of the fully resolved path to the logical key.
+                        prepend = hashlib.sha256(str(pk).encode("utf-8")).hexdigest()[:8]
+                        lk = f"{col_label}/{prepend}_{val.name}"
                         if pk.is_file():
-                            key = str(uuid4()).replace("-", "")
-                            unique_name = f"{key}_{val.name}"
-                            lk = f"{col_label}/{unique_name}"
                             v_ds.data[col].values[i] = lk
 
-                            # Attach metadata to object
-                            meta = {index_col: str(v_ds.data[index_col].values[i]) for index_col in self.index_columns}
-                            pkg.set(lk, pk, meta)
+                            # Create metadata dictionary to attach to object
+                            meta = {}
+                            for index_col in self.index_columns:
```

Comment:
The cyclomatic complexity of this function seems pretty high - can the logic here be broken up into multiple pieces?

Reply:
Yea.... There are now a couple of functions in this repo with a cyclomatic complexity of ~20.... will try to reduce.